src/ctap2/commands/mod.rs

```diff
@@ -92,66 +95,179 @@ pub trait RequestCtap2: fmt::Debug {
+    UsingInternalUv,
+    /// Successfully established PinUvAuthToken via GetPinToken (CTAP2.0)
+    SuccessGetPinToken,
+}
+
+pub(crate) trait PinUvAuthCommand: RequestCtap2 {
     fn pin(&self) -> &Option<Pin>;
     fn set_pin(&mut self, pin: Option<Pin>);
-    fn pin_auth(&self) -> &Option<PinUvAuthParam>;
     fn set_pin_auth(&mut self, pin_auth: Option<PinUvAuthParam>);
     fn client_data_hash(&self) -> ClientDataHash;
     fn unset_uv_option(&mut self);
-    fn determine_pin_auth<D: FidoDevice>(&mut self, dev: &mut D) -> Result<(), AuthenticatorError> {
+
+    /// If Ok() is returned, the request can be sent to the device
+    /// If Err() is returned, either cancel the operation or ask the user for a PIN
+    fn determine_pin_uv_auth<D: FidoDevice>(
+        &mut self,
+        dev: &mut D,
+        skip_uv: bool,
+    ) -> Result<PinUvAuthResult, AuthenticatorError> {
+        // CTAP1/U2F-only devices do not support PinUvAuth, so we skip it
         if !dev.supports_ctap2() {
             self.set_pin_auth(None);
-            return Ok(());
+            return Ok(PinUvAuthResult::DeviceIsCtap1);
         }
 
-        let client_data_hash = self.client_data_hash();
-        let pin_auth = calculate_pin_auth(dev, &client_data_hash, self.pin())
+        self.determine_pin_uv_auth_ctap_2_0(dev, skip_uv)
+    }
+
+    /// CTAP 2.0-only version:
+    /// "Getting pinUvAuthToken using getPinToken (superseded)"
+    fn determine_pin_uv_auth_ctap_2_0<D: FidoDevice>(
+        &mut self,
+        dev: &mut D,
+        skip_uv: bool,
+    ) -> Result<PinUvAuthResult, AuthenticatorError> {
+        let info = dev
+            .get_authenticator_info()
+            .ok_or(AuthenticatorError::Platform)?;
+
+        if self.pin().is_none() {
+            // Only use UV, if the device supports it and we don't skip it
+            // which happens as a fallback, if UV-usage failed too many times
+            let supports_uv = !skip_uv && info.options.user_verification == Some(true);
+            let supports_pin = info.options.client_pin.is_some();
+            let pin_configured = info.options.client_pin == Some(true);
+
+            // If the device supports internal user-verification (e.g. fingerprints),
+            // skip PIN-stuff
+            if supports_uv {
+                return Ok(PinUvAuthResult::UsingInternalUv);
+            }
+
+            // Device does not support any auth-method
+            if !supports_pin && !supports_uv {
+                // We'll send it to the device anyways, and let it error out (or magically work)
+                return Ok(PinUvAuthResult::NoAuthTypeSupported);
+            }
+
+            // Device supports PIN: Ask user for it.
+            if !supports_uv && pin_configured {
+                return Err(AuthenticatorError::PinError(PinError::PinRequired));
+            }
+        }
+
+        let pin_auth = dev
+            .get_pin_token(&self.client_data_hash(), self.pin())
```

Comment:
This will ultimately call `set_shared_secret` on `Device`. Where are we doing that in the `UsingInternalUv` case?

Maybe we need to implement and integrate `getPinUvAuthTokenUsingUvWithPermissions` before we can land this patch?

Reply:
Ah, you are right. I was under the impression, that we do not need the shared_secret, if we don't use a PIN. But we may need it for the hmac-extension (which I'm currently not sure, if it works correctly anyways).
I'll add a `dev.establish_shared_secret()?;` to the UsingInternalUv-case.

Waiting for `getPinUvAuthTokenUsingUvWithPermissions` will (probably) not change anything here, since the plan is to do that in a separate function (hence the name `determine_pin_uv_auth_ctap_2_0()`). 
The InternalUv-case needs to work for CTAP2.0-only tokens as well.

Reply:
OK, sounds good. I didn't realize that Internal UV did not require the shared secret in 2.0, but we do need it for the hmac extension (even though it's not currently working).